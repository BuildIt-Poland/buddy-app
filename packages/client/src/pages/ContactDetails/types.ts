import { RouteComponentProps } from 'react-router-dom';
import {
  QueryNewbieArgs,
  QueryBuddyArgs,
  QueryTalentArgs,
  User,
} from '@buddy-app/schema';

export type BasicDetailsParams = QueryNewbieArgs | QueryBuddyArgs | QueryTalentArgs;

export type UserBasicDetails = {
  [key: string]: User;
};

export type ContactDetailsProps = Pick<RouteComponentProps, 'history'>;
