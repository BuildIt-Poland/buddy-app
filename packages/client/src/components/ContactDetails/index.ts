import ContactDetails from './ContactDetails';

export default ContactDetails;
