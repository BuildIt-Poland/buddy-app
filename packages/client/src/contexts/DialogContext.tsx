import React, { useState, useContext } from 'react';
import AlertDialog from 'atoms/AlertDialog';

interface State {
  isOpen: boolean;
  message: string;
  title: string | undefined;
  onConfirm?: () => void;
}

interface DialogContextData {
  showDialog: (message: string, title?: string, onConfirm?: () => void) => void;
  hideDialog: () => void;
}

const defaultState: State = {
  isOpen: false,
  title: '',
  message: '',
};

const DialogContext = React.createContext<DialogContextData | undefined>(undefined);

interface DialogProviderProps {
  children: React.ReactNode;
  value?: DialogContextData;
}

const DialogProvider = (props: DialogProviderProps): JSX.Element => {
  const [state, setDialog] = useState<State>(defaultState);

  const showDialog = (message: string, title?: string, onConfirm?: () => void) =>
    setDialog({ isOpen: true, message, title, onConfirm });

  const hideDialog = () => setDialog(defaultState);

  const onConfirmHandler = () => {
    state.onConfirm && state.onConfirm();
    hideDialog();
  };

  return (
    <DialogContext.Provider
      value={{
        showDialog,
        hideDialog,
      }}>
      <AlertDialog
        isOpen={state.isOpen}
        message={state.message}
        onConfirm={state.onConfirm ? onConfirmHandler : undefined}
        title={state.title}
        onClose={hideDialog}
      />
      {props.children}
    </DialogContext.Provider>
  );
};

const useDialog = () => {
  const context = useContext(DialogContext);
  if (context === undefined) {
    throw new Error(`useDialog must be used within a DialogProvider`);
  }
  return context;
};

export { DialogProvider, useDialog };
