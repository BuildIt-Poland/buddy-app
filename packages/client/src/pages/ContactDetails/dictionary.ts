interface Dictionary {
  TITLE: string;
  EDIT_BUTTON_TITLE: string;
}

const DICTIONARY: Dictionary = {
  TITLE: 'Contact Details',
  EDIT_BUTTON_TITLE: 'Update Contact Details',
};

export default DICTIONARY;
