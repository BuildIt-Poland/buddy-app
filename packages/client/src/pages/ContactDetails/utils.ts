import { UserRole } from '@buddy-app/schema';
import { ROUTES } from 'shared/routes';
import {
  NEWBIE_CONTACT_DETAILS,
  BUDDY_CONTACT_DETAILS,
  TALENT_CONTACT_DETAILS,
} from 'graphql/contact-details.graphql';

export const getQueryData = (userId: string, newbieId: string, buddyId: string) =>
  !newbieId && !buddyId
    ? {
        [UserRole.Newbie]: {
          query: NEWBIE_CONTACT_DETAILS,
          variables: { newbieId: userId },
          userRole: UserRole.Newbie,
          editRoute: ROUTES.NEWBIE_EDIT_DETAILS,
        },
        [UserRole.Buddy]: {
          query: BUDDY_CONTACT_DETAILS,
          variables: { buddyId: userId },
          userRole: UserRole.Buddy,
          editRoute: ROUTES.BUDDY_EDIT_DETAILS,
        },
        [UserRole.Talent]: {
          query: TALENT_CONTACT_DETAILS,
          variables: { talentId: userId },
          userRole: UserRole.Talent,
          editRoute: ROUTES.TALENT_EDIT_DETAILS,
        },
      }
    : {
        [UserRole.Newbie]: {
          query: BUDDY_CONTACT_DETAILS,
          variables: { buddyId },
          userRole: UserRole.Buddy,
        },
        [UserRole.Buddy]: {
          query: NEWBIE_CONTACT_DETAILS,
          variables: { newbieId },
          userRole: UserRole.Newbie,
          editRoute: ROUTES.BUDDY_EDIT_NEWBIE_DETAILS.replace(':newbieId', newbieId),
        },
        [UserRole.Talent]: {
          query: newbieId ? NEWBIE_CONTACT_DETAILS : BUDDY_CONTACT_DETAILS,
          variables: { newbieId, buddyId },
          userRole: newbieId ? UserRole.Newbie : UserRole.Buddy,
          editRoute: newbieId
            ? ROUTES.TALENT_EDIT_NEWBIE_DETAILS.replace(':newbieId', newbieId)
            : ROUTES.TALENT_EDIT_BUDDY_DETAILS.replace(':buddyId', buddyId),
        },
      };
