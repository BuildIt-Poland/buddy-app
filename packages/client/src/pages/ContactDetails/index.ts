import ContactDetails from './ContactDetails';
export * from './utils';
export * from './types';

export default ContactDetails;
