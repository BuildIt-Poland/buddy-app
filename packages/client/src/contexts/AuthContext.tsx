import React, { useReducer, useEffect, useContext } from 'react';
import { Mutation, AuthPayload } from '@buddy-app/schema';
import { useQuery } from '@apollo/react-hooks';
import { ExecutionResult } from '@apollo/react-common';
import { auth } from 'utils';
import { LOGIN_MUTATION } from 'graphql/login.graphql';
import { GET_LOCAL_AUTH } from 'graphql/get-auth.graphql';
import { AuthCache, setCacheToken } from 'utils/apollo-client/cache';
import { apolloClient } from 'utils';

enum ActionTypes {
  AUTH_INIT = 'auth/init',
  AUTH_ERROR = 'auth/error',
  AUTH_SUCCESS = 'auth/success',
  AUTH_LOGOUT = 'auth/logout',
}

interface Action {
  type: ActionTypes;
  payload?: AuthPayload;
}

export interface AuthState {
  loading: boolean;
  error?: any;
  data: AuthPayload;
  isAuthenticated: boolean;
}

const defaultState: AuthState = {
  loading: false,
  error: null,
  data: {} as AuthPayload,
  isAuthenticated: false,
};

interface AuthContext extends AuthState {
  login: (email: string, password: string) => Promise<ExecutionResult<Mutation>>;
  logout: () => void;
}

interface AuthProviderProps {
  children: React.ReactNode;
  value?: AuthContext;
}

const defaultContext: AuthContext = {
  ...defaultState,
  login: () => Promise.reject(),
  logout: () => null,
};

const AuthContext = React.createContext(defaultContext);

const authReducer = (state: AuthState, action: Action): AuthState => {
  switch (action.type) {
    case ActionTypes.AUTH_INIT:
      return {
        ...state,
        loading: true,
        error: null,
      };
    case ActionTypes.AUTH_SUCCESS:
      return {
        ...state,
        loading: false,
        data: { ...action.payload } as AuthPayload,
        isAuthenticated: true,
      };
    case ActionTypes.AUTH_ERROR:
      return {
        ...state,
        loading: false,
        error: { ...action.payload },
      };
    case ActionTypes.AUTH_LOGOUT:
      return {
        ...state,
        data: {} as AuthPayload,
        isAuthenticated: false,
      };
    default:
      throw new Error('Not valid action');
  }
};

const AuthProvider = ({ children, ...props }: AuthProviderProps) => {
  const [state, dispatch] = useReducer(authReducer, defaultState);
  const { data } = useQuery<AuthCache>(GET_LOCAL_AUTH);

  useEffect(() => {
    const user = auth.getUser();
    if (user) {
      dispatch({
        type: ActionTypes.AUTH_SUCCESS,
        payload: user,
      });
    }
  }, []);

  useEffect(() => {
    if (data && data.auth && data.auth.tokenHasExpired) {
      logout();
    }
  }, [data]);

  const login = async (email: string, password: string) => {
    dispatch({ type: ActionTypes.AUTH_INIT });

    try {
      const { data } = await apolloClient.mutate<Mutation>({
        mutation: LOGIN_MUTATION,
        variables: {
          email,
          password,
        },
      });

      if (data) {
        auth.setUser(data.login);
        dispatch({
          type: ActionTypes.AUTH_SUCCESS,
          payload: data.login,
        });
      }

      return Promise.resolve({ data });
    } catch (error) {
      dispatch({ type: ActionTypes.AUTH_ERROR, payload: error });
      return Promise.reject(error);
    }
  };

  const logout = () => {
    auth.removeUser();
    apolloClient.writeData({
      data: {
        ...setCacheToken(false),
      },
    });
    dispatch({ type: ActionTypes.AUTH_LOGOUT });
  };

  return (
    <AuthContext.Provider value={{ ...state, login, logout }} {...props}>
      {children}
    </AuthContext.Provider>
  );
};

const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error(`useAuth must be used within a AuthProvider`);
  }
  return context;
};

export { AuthProvider, useAuth };
