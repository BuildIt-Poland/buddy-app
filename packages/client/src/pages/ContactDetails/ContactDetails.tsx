import React from 'react';
import { useParams, Link } from 'react-router-dom';
import { useQuery } from '@apollo/react-hooks';
import {
  NEWBIE_CONTACT_DETAILS,
  BUDDY_CONTACT_DETAILS,
  TALENT_CONTACT_DETAILS,
} from 'graphql/contact-details.graphql';
import { QueryBuddyArgs, QueryNewbieArgs, UserRole } from '@buddy-app/schema';
import { useAuth } from 'contexts/AuthContext';
import UserDetails from 'components/UserDetails';
import { BasicDetailsParams, UserBasicDetails } from 'components/UserMenu/types';
import BackPageContainer from 'atoms/BackPageContainer';
import ContactDetailsPlaceHolder from 'atoms/ContactDetailsPlaceHolder';
import EditButton from 'atoms/EditButton';
import { ROUTES } from 'shared/routes';
import { ContactDetailsProps } from './types';
import DICTIONARY from './dictionary';

const ContactDetails: React.FC<ContactDetailsProps> = ({ history }) => {
  const { newbieId, buddyId } = useParams<QueryNewbieArgs & QueryBuddyArgs>();
  const isCurrentUserDetails = !newbieId && !buddyId;
  const {
    data: { role, userId },
  } = useAuth();

  const queryByRole = isCurrentUserDetails
    ? {
        [UserRole.Newbie]: {
          query: NEWBIE_CONTACT_DETAILS,
          variables: { newbieId: userId },
          userRole: UserRole.Newbie.toLowerCase(),
          editRoute: ROUTES.NEWBIE_EDIT_DETAILS,
        },
        [UserRole.Buddy]: {
          query: BUDDY_CONTACT_DETAILS,
          variables: { buddyId: userId },
          userRole: UserRole.Buddy.toLowerCase(),
          editRoute: ROUTES.BUDDY_EDIT_DETAILS,
        },
        [UserRole.Talent]: {
          query: TALENT_CONTACT_DETAILS,
          variables: { talentId: userId },
          userRole: UserRole.Talent.toLowerCase(),
          editRoute: ROUTES.TALENT_EDIT_DETAILS,
        },
      }
    : {
        [UserRole.Newbie]: {
          query: BUDDY_CONTACT_DETAILS,
          variables: { buddyId },
          userRole: UserRole.Buddy.toLowerCase(),
        },
        [UserRole.Buddy]: {
          query: NEWBIE_CONTACT_DETAILS,
          variables: { newbieId },
          userRole: UserRole.Newbie.toLowerCase(),
          editRoute: ROUTES.BUDDY_EDIT_NEWBIE_DETAILS.replace(':newbieId', newbieId),
        },
        [UserRole.Talent]: {
          query: newbieId ? NEWBIE_CONTACT_DETAILS : BUDDY_CONTACT_DETAILS,
          variables: { newbieId, buddyId },
          userRole: newbieId
            ? UserRole.Newbie.toLowerCase()
            : UserRole.Buddy.toLowerCase(),
          editRoute: newbieId
            ? ROUTES.TALENT_EDIT_NEWBIE_DETAILS.replace(':newbieId', newbieId)
            : ROUTES.TALENT_EDIT_BUDDY_DETAILS.replace(':buddyId', buddyId),
        },
      };

  const { query, variables, userRole, editRoute } = queryByRole[role];
  const { data, loading } = useQuery<UserBasicDetails, BasicDetailsParams>(query, {
    variables,
  });
  const userDetails = data && data[userRole as string];

  return (
    <BackPageContainer
      title={DICTIONARY.TITLE}
      id='contact-details-page'
      backGroundShape>
      {loading && <ContactDetailsPlaceHolder />}
      {userDetails && <UserDetails details={userDetails} />}
      {editRoute && (
        <Link to={editRoute}>
          <EditButton
            aria-label='Edit contact details'
            title={DICTIONARY.EDIT_BUTTON_TITLE}
          />
        </Link>
      )}
    </BackPageContainer>
  );
};

export default ContactDetails;
