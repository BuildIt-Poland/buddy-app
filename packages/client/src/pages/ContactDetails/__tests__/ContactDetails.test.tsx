import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { MockedProvider } from '@apollo/react-testing';
import { act, create } from 'react-test-renderer';
import waitForExpect from 'wait-for-expect';
import ContactDetails from 'pages/ContactDetails';
import {
  buddyContactDetails,
  newbieContactDetails,
  mockedBuddyContext,
  mockedNewbieContext,
} from '__mocks__';
import { AuthProvider } from 'contexts/AuthContext';

jest.mock('atoms/EditButton', () => 'EditButton');
jest.mock('components/UserDetails', () => 'UserDetails');
jest.mock('atoms/ContactDetailsPlaceHolder', () => 'ContactDetailsPlaceHolder');
jest.mock('atoms/BackPageContainer', () => 'BackPageContainer');

describe('Component - ContactDetails', () => {
  const mockHistory: any = {
    push: jest.fn(),
  };

  describe('when logged in as buddy', () => {
    const path = '/buddy/newbies/1234/details';
    const component = create(
      <MockedProvider
        mocks={newbieContactDetails({ newbieId: '1234' })}
        addTypename={false}
        resolvers={{}}>
        <MemoryRouter initialEntries={[path]}>
          <Route path={'/buddy/newbies/:newbieId/details'}>
            <AuthProvider value={mockedBuddyContext()}>
              <ContactDetails history={mockHistory} />
            </AuthProvider>
          </Route>
        </MemoryRouter>
      </MockedProvider>
    );
    it('renders correctly', async () => {
      expect(component).toMatchSnapshot();

      await act(async () => {
        await waitForExpect(() => {
          expect(component).toMatchSnapshot();
        });
      });
    });
  });

  describe('when logged in as newbie', () => {
    const path = '/newbie/buddy/1234/details';
    const component = create(
      <MockedProvider
        mocks={buddyContactDetails({ buddyId: '1234' })}
        addTypename={false}
        resolvers={{}}>
        <MemoryRouter initialEntries={[path]}>
          <Route path={'/newbie/buddy/:buddyId/details'}>
            <AuthProvider value={mockedNewbieContext()}>
              <ContactDetails history={mockHistory} />
            </AuthProvider>
          </Route>
        </MemoryRouter>
      </MockedProvider>
    );
    it('renders correctly', async () => {
      expect(component).toMatchSnapshot();

      await act(async () => {
        await waitForExpect(() => {
          expect(component).toMatchSnapshot();
        });
      });
    });
  });
});
