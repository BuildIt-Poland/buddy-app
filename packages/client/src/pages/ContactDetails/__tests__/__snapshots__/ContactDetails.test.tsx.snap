// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - ContactDetails when logged in as buddy renders correctly 1`] = `
<BackPageContainer
  backGroundShape={true}
  id="contact-details-page"
  title="Contact Details"
>
  <ContactDetailsPlaceHolder />
  <a
    href="/buddy/newbies/1234/edit-details"
    onClick={[Function]}
  >
    <EditButton
      aria-label="Edit contact details"
      disabled={false}
      title="Update Contact Details"
    />
  </a>
</BackPageContainer>
`;

exports[`Component - ContactDetails when logged in as buddy renders correctly 2`] = `
<BackPageContainer
  backGroundShape={true}
  id="contact-details-page"
  title="Contact Details"
>
  <UserDetails
    details={
      Object {
        "email": "tom@wipro.com",
        "id": "1234",
        "name": "Tom Hanks",
        "notes": "Nice guy",
        "phoneNumber": "1234567",
        "photo": "some-url",
        "position": "Dev Ops",
        "role": "NEWBIE",
        "startDate": "2010-01-01",
      }
    }
  />
  <a
    href="/buddy/newbies/1234/edit-details"
    onClick={[Function]}
  >
    <EditButton
      aria-label="Edit contact details"
      disabled={false}
      title="Update Contact Details"
    />
  </a>
</BackPageContainer>
`;

exports[`Component - ContactDetails when logged in as newbie renders correctly 1`] = `
<BackPageContainer
  backGroundShape={true}
  id="contact-details-page"
  title="Contact Details"
>
  <UserDetails
    details={
      Object {
        "email": "tom@wipro.com",
        "id": "1234",
        "name": "Tom Hanks",
        "phoneNumber": "1234567",
        "photo": "some-url",
        "position": "Dev Ops",
        "role": "BUDDY",
        "startDate": "2010-01-01",
      }
    }
  />
</BackPageContainer>
`;

exports[`Component - ContactDetails when logged in as newbie renders correctly 2`] = `
<BackPageContainer
  backGroundShape={true}
  id="contact-details-page"
  title="Contact Details"
>
  <UserDetails
    details={
      Object {
        "email": "tom@wipro.com",
        "id": "1234",
        "name": "Tom Hanks",
        "phoneNumber": "1234567",
        "photo": "some-url",
        "position": "Dev Ops",
        "role": "BUDDY",
        "startDate": "2010-01-01",
      }
    }
  />
</BackPageContainer>
`;
